drivers/stepper/gpio_stepper_controller.c

```diff
@@ -0,0 +1,231 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024 Carl Zeiss Meditec AG
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT gpio_stepper
+
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys_clock.h>
+#include <zephyr/drivers/stepper.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/__assert.h>
+
+LOG_MODULE_REGISTER(gpio_stepper_motor_controller, CONFIG_STEPPER_LOG_LEVEL);
+
+#define NUMBER_OF_GPIO_PINS 4
+
+struct gpio_stepper_motor_controller_config {
+	const struct gpio_dt_spec *control_pins;
+};
+
+struct gpio_stepper_motor_controller_data {
+	const struct device *dev;
+	struct k_spinlock lock;
+	struct k_work_delayable stepper_dwork;
+	struct k_poll_signal *async_signal;
+	int32_t actual_position;
+	int32_t target_position;
+	uint32_t delay_in_us;
+};
+
+static int stepper_motor_move_with_4_gpios(const struct gpio_dt_spec *gpio_spec,
+					   int32_t step_number)
+{
+	switch (step_number) {
+	case 0:
+		LOG_DBG("Step 1010");
+		gpio_pin_configure_dt(&gpio_spec[0], GPIO_OUTPUT_ACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[1], GPIO_OUTPUT_INACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[2], GPIO_OUTPUT_ACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[3], GPIO_OUTPUT_INACTIVE);
+		break;
+
+	case 1:
+		LOG_DBG("Step 0110");
+		gpio_pin_configure_dt(&gpio_spec[0], GPIO_OUTPUT_INACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[1], GPIO_OUTPUT_ACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[2], GPIO_OUTPUT_ACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[3], GPIO_OUTPUT_INACTIVE);
+		break;
+
+	case 2:
+		LOG_DBG("Step 0101");
+		gpio_pin_configure_dt(&gpio_spec[0], GPIO_OUTPUT_INACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[1], GPIO_OUTPUT_ACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[2], GPIO_OUTPUT_INACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[3], GPIO_OUTPUT_ACTIVE);
+		break;
+
+	case 3:
+		LOG_DBG("Step 1001");
+		gpio_pin_configure_dt(&gpio_spec[0], GPIO_OUTPUT_ACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[1], GPIO_OUTPUT_INACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[2], GPIO_OUTPUT_INACTIVE);
+		gpio_pin_configure_dt(&gpio_spec[3], GPIO_OUTPUT_ACTIVE);
+		break;
+	}
+	return 0;
+}
+
+void stepper_work_step_handler(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct gpio_stepper_motor_controller_data *data =
+		CONTAINER_OF(dwork, struct gpio_stepper_motor_controller_data, stepper_dwork);
+	const struct gpio_stepper_motor_controller_config *config = data->dev->config;
+
+	static int32_t step_count = 1;
```

Comment:
avoid static variables on function level – bad code style – BTW, why you introduce a separate motor step counter – use one central counter in drivers data area